### DISCUSSION TITLE
How to enable C# 8 `Nullable` by `<PropertyGroup Condtion='xxx'>`?

### DISCUSSION THREAD

**Discussion by GF-Huang** [dotnet/roslyn#64563] [Q&A]
My library targets to multiple frameworks: `net462;netcoreapp2.1;netcoreapp2.2;netcoreapp3.0;netcoreapp3.1;net5.0;net6.0`.

And I want to enable C# 8 reference type nullable, but it's only available above .NET Core 3.x.

So I use the following:
```xml
<PropertyGroup Condition="$([MSBuild]::VersionGreaterThanOrEquals($(LangVersion),'8.0'))">
    <Nullable>enable</Nullable>
</PropertyGroup>
```

`[MSBuild]::VersionGreaterThanOrEquals` is the [MSBuild version-comparison functions](https://learn.microsoft.com/en-us/visualstudio/msbuild/property-functions?view=vs-2022#msbuild-version-comparison-functions).

But I got error when I save the `*.csproj`:

![image](https://user-images.githubusercontent.com/4510984/194448777-9b892249-1271-48ba-9c1c-2a7f9942a3c9.png)

This means the `$(LangVersion)` part got empty, but if I print `$(LangVersion)` at build time, it's not empty.

![image](https://user-images.githubusercontent.com/4510984/194448982-c7b43981-f4d4-498a-a7d1-5da9d7be1dc2.png)

When I comment the error part, I can save my `*.csproj` file, so I can build and see the print message of `$(LangVersion)` now.

![image](https://user-images.githubusercontent.com/4510984/194450365-6b9fe86c-c5ab-4a72-8002-3d7c07fb3588.png)

There is an empty at the last, it's probably the culprit. Why? What is this? How to solve this?

**Comment by RikkiGibson** [comment]
There could be fairly complex reasons why this happens, related to the cyclic nature of needing to evaluate a condition to set a property value, but also needing to know the property values in order to evaluate the condition.

One solution you might want to try is to use Nullable "annotations" mode in all your target frameworks, and only do the *warnings* side of it in your newest target framework. For example:

```cs
<Nullable Condition="$(TargetFramework) == 'net6.0'">enable</Nullable>
<Nullable Condition="$(Nullable) == ''">annotations</Nullable>
```

See https://learn.microsoft.com/en-us/dotnet/csharp/nullable-references#nullable-contexts for more info.

**Reply by GF-Huang** [reply]
Thanks your reply, but I had tried your code, get error at build time due to **nullable unavailable at C# 7.3**.

Syntax error below .NET Core 3.x.

![image](https://user-images.githubusercontent.com/4510984/195269884-d1ab2fdd-6d3b-4274-9a8b-05bf11dadbff.png)


![image](https://user-images.githubusercontent.com/4510984/195269667-426c689e-cc54-461d-ae94-9773d456e6a8.png)

**Reply by huoyaoyuan** [reply]
You can just set `LangVersion` to 8, or even 10.

**Reply by GF-Huang** [reply]
> You can just set `LangVersion` to 8, or even 10.

Solved by https://github.com/manuelroemer/Nullable, do you think this is a good approach?

**Reply by huoyaoyuan** [reply]
It should be OK, although most projects just add the attributes themselves.

**Reply by GF-Huang** [reply]
> add the attributes themselves

Could you explain what this mean?

**Reply by huoyaoyuan** [reply]
The attributes like `[MaybeNullWhen]` are a part of C# 9, and only defined in .NET 5+.
If you just use the C# 8 rules, you don't need the package at all.

**Reply by taori** [reply]
@GF-Huang RikkiGibson solved this, no?

**Reply by GF-Huang** [reply]
> @GF-Huang RikkiGibson solved this, no?

No.

**Reply by taori** [reply]
have you tried this?
```
<PropertyGroup Condition=" '$(TargetFramework)' == 'net6.0' ">
    <Nullable>enable</Nullable>
</PropertyGroup>
```

**Reply by GF-Huang** [reply]
> have you tried this?
> 
> ```
> <PropertyGroup Condition=" '$(TargetFramework)' == 'net6.0' ">
>     <Nullable>enable</Nullable>
> </PropertyGroup>
> ```

I don't want to specific every `TargetFramework` manual, so I need a `>` or `>=` or `<` or `<=` operator condition to enable `Nullable`.

**Comment by dnperfors** [comment]
Why not just enable it for all versions? When compiling with the latest SDk and the latest language version, you will get the correct result even on .NET Framework.
I do the exact same thing on my own library and it works like a charm.

**Reply by taori** [reply]
Nullable would spawn hundreds of warnings regarding nullability if a product was developed for years without taking care of every place that has nullable concerns. Pretty sure that is why.

More importantly imagine maybe a couple dozen very important warnings disappearing among hundreds of warnings introduced by an altrication on how to properly code around null and appending ? to every type. I think most people who at some point enabled nullable are happily they did not get thrown to warning hell by someone elses decision just because they wanted to use (other) new language features.

**Reply by dnperfors** [reply]
That makes sense for the code that uses the library. Enabling or disabling nullability inside a single project based on the language version just doesn't make sense to me. Either way, if you really want this, it is probably better to art the value based on the framework
